server/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java

```diff
@@ -379,19 +378,22 @@ void removeTombstoneUnderLock(BytesRef uid) {
     void pruneTombstones(long currentTime, long pruneInterval) {
         for (Map.Entry<BytesRef, DeleteVersionValue> entry : tombstones.entrySet()) {
             BytesRef uid = entry.getKey();
-            try (Releasable ignored = acquireLock(uid)) { // can we do it without this lock on each value? maybe batch to a set and get
-                // the lock once per set?
-                // Must re-get it here, vs using entry.getValue(), in case the uid was indexed/deleted since we pulled the iterator:
-                DeleteVersionValue versionValue = tombstones.get(uid);
-                if (versionValue != null) {
-                    // check if the value is old enough to be removed
-                    final boolean isTooOld = currentTime - versionValue.time > pruneInterval;
-                    if (isTooOld) {
-                        // version value can't be removed it's
-                        // not yet flushed to lucene ie. it's part of this current maps object
-                        final boolean isNotTrackedByCurrentMaps = versionValue.time < maps.getMinDeleteTimestamp();
-                        if (isNotTrackedByCurrentMaps) {
-                            removeTombstoneUnderLock(uid);
+            Releasable lock = keyedLock.tryAcquire(uid);
```

Comment:
I have to admit that it took me a moment to understand *why* this is avoiding the deadlock as there need to be two locks involved that are acquired by two threads in opposite order. The reason why this fix works is that this is the inner of the two involved locks. Then there are two cases:

1. We get this lock, remove the tombstone and return the lock. If another thread is trying to acquire this lock, it can do so after we have left the try-with-resources block.
2. We do not get the lock. Then we will not wait but rather give up (this time).

(just wrote this down for my own reference.)

Reply:
correct